xds/src/main/java/com/linecorp/armeria/xds/AbstractResourceNode.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.xds;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.envoyproxy.envoy.config.core.v3.ConfigSource;
+import io.grpc.Status;
+
+abstract class AbstractResourceNode<T> implements ResourceNode<AbstractResourceHolder> {
```

Comment:
>  Since this requires a lot of changes, let me try this after this PR is merged if it makes sense.

It makes sense, but you probably don't want `AbstractResourceNode` since inheritance doesn't bode well with generic types. (So you would probably define a single `DefaultResourceNode<T>`)

Also, the map implementation in `SubscriberStorage` may become tricky. (we probably want the entire call path up to `ResourceParser` to be type-safe if we want this to be meaningful)

I've already tried and reverted this approach, but feel free to try it again

Reply:
Got your point. Let me try that later after we merge this. 😉